core/src/main/scala/kafka/controller/ControllerContext.scala

```diff
@@ -61,7 +61,7 @@ case class ReplicaAssignment(replicas: Seq[Int],
 class ControllerContext {
   val stats = new ControllerStats
   var offlinePartitionCount = 0
-  var shuttingDownBrokerIds: mutable.Set[Int] = mutable.Set.empty
+  var shuttingDownBrokerIds: Map[Int, Long] = Map.empty
```

Comment:
Currently we only maintain the current epoch of each broker in the liveBrokerEpochs.
This PR suggests the epoch in this map may become lagging behind the epochs maintained in the liveBrokerEpochs. Can you please give an example sequence of events that can cause that condition?

Reply:
This map just contains the epoch of each broker when it had last initialized shutdown via the controlled shutdown protocol. So it's OK for it to lag behind the actual epoch of live brokers.

The idea is that ControllerContext#liveBrokerIds will contain the set of broker ids that are considered live. A broker that has been allowed to initiate shutdown is not considered live for the uses of this function.

It is defined as:

```
def liveBrokerIds: Set[Int] = liveBrokerEpochs.filter(b => b._2 > (shuttingDownBrokerIds.getOrElse(b._1, -1): Long)).keySet
```

So take the (brokerId, epoch) for every known live broker, and filter out all brokers which have been marked as shutting down with that brokerEpoch.

So say you start with brokerId 3, epoch 12:

  liveBrokerEpochs contains 3 -> 12
  shuttindDownBrokerIds is empty

If broker 3 is allowed to shut down, then shuttingDownBrokerIds has 3 -> 12 added to it. Now the state is


  liveBrokerEpochs contains 3 -> 12
  shuttingDownBrokerIds contains 3 -> 12

With that state, liveBrokerIds evaluates to the empty set.

When broker 3 comes back up with epoch 13 the state will be 

  liveBrokerEpochs contains 3 -> 13
  shuttingDownBrokerIds contains 3 -> 12

Now since shuttingDownBrokerIds has a the epoch for the previous broker process with id 3, liveBrokerIds will evaluate to the set `{3}`

So it's fine for shuttingDownBrokerIds to have old broker ids. It is just a marker to inform the cluster the epoch at which a given broker had initiated shutdown.

Reply:
Thanks for the example. It's still not clear to me why we need to keep track of the epoch of when a broker initiated the controlled shutdown. What would be missing if we don't?

Reply:
The missing detail is controller movement.

The existing controlled shutdown logic before this patch doesn't keep track of the shutdown epoch because the only reason it cares about the epoch at all is to filter out stale ControlledShutdownRequests that come in for a broker that has already been restarted.

```
    // broker epoch in the request is unknown if the controller hasn't been upgraded to use KIP-380
    // so we will keep the previous behavior and don't reject the request
    if (brokerEpoch != AbstractControlRequest.UNKNOWN_BROKER_EPOCH) {
      val cachedBrokerEpoch = controllerContext.liveBrokerIdAndEpochs(id)
      if (brokerEpoch < cachedBrokerEpoch) {
        val stateBrokerEpochErrorMessage = "Received controlled shutdown request from an old broker epoch " +
          s"$brokerEpoch for broker $id. Current broker epoch is $cachedBrokerEpoch."
        info(stateBrokerEpochErrorMessage)
        throw new StaleBrokerEpochException(stateBrokerEpochErrorMessage)
      }
    }
```

After that block it no longer cares about the epoch and discards that information, only storing the broker id of a shutting down broker into controllerContext.shuttingDownBrokers, which is a Set[Int].

After this patch, the controller is keeping track of which brokers have initiated shutdown so that it doesn't allow multiple brokers to shut down that will have an impact on the availability of a topic-partition they have in common. So it needs a short term history of shutdown requests.

The existing shuttingDownBrokers: Set[Int] could work fine for this, however it doesn't deal with controller movement.

Imagine this hypothetical scenario:
1. You have controller running on broker id 0
1. You have two brokers 2 and 3, and a topic-partition topic-0 with ISR [1, 2].
2. Broker id 2 sends ControlledShutdownRequest to controller 0. Controller does safety check and says 👍 go ahead and shut down.
3. Broker 2 gets ControlledShutdownResponse and begins shutdown
4. Broker 0 crashes and controller moves to broker 1
5. Broker 3 sends ControlledShutdownREquest to controller 1.
6. Controller 1 has no idea that broker 2 is shutting down. Controller does safety check and says 👍  go ahead broker 3, shut down
7. Broker 2 shuts down
8. Broker 3 shuts down, leaving topic-0 offline.

So, that means that in order to handle controller movement, you need the controller to pass state along to the next controller so that it knows which brokers are pending shutdown.

You could do that just by writing the broker id of the brokers pending shutdown to zookeeper, effectively persisting shuttingDownBrokers: Set[Int] to zookeeper.

The problem is, when a new controller boots up after a controlelr move, then it would only have the broker ids of brokers who had initiated shutdown, without knowing whether that broker had already shut down and started back up again.

Storing brokerId => brokerEpoch in zookeeper under /brokers/shutdown means that when the new controller takes over, it will know right away whether that state passed from the previous controller is stale shutdown information.